Base/src/main/java/io/deephaven/base/ringbuffer/CharRingBuffer.java

```diff
@@ -121,6 +135,7 @@ public int capacity() {
     }
 
     public int remaining() {
+        maybeFixIndices();
```

Comment:
I don't personally much like all these calls to `maybeFixIndices()` sprinkled throughout the code (5 different places by my count). It makes it difficult for future readers/maintainers to have confidence exactly what code needs a call to `maybeFixIndices`. For example they may wonder (as I do) why a non-mutating method like `remaining()` needs a fix-indices check. In my opinion the answer to all these questions is "the only code that might need fix-indices is code that changes `tail`". Along those lines the check needs to be in only one place: inside `addUnsafe`. I realize that doing so makes `addUnsafe` look longer and less tight, but I want to double- and quadruple- down on the argument that highly-predictable branches are basically free. "Most" user code is going to do the wrap check somewhere anyway, so from a performance standpoint it's moot whether the tail got checked inside `add()` or `addUnsafe()`. For code where it does matter (some hypothetical tight loop that doesn't call `add()`, only `addUnsafe()`), I think we can rely on the branch predictor to smoothly skip over the cost of the `if`.

If you're really unconvinced, you could also run your (existing! not new) benchmarks and/or tests with and without the wrap check to confirm that there is no wall-clock-observable difference between the presence and absence of the wrap check. Since we know that the wrap check is only a theoretical hazard anyway, we could run our code with no checks at all to see if it's any faster.

If you do buy all of the above, you could inline maybeFixIndices() inside addUnsafe() rather than being a separate function call to make it feel a little tighter.

Reply:
I'm convinced.